File: packages/agents/score.py
"""
Agente de Qualificação — motor de lead score.

Implementa a régua de pontuação da seção 6.3 do prompt-mestre:

  product fit             0-20
  marketplace gap         0-20
  business structure      0-15
  catalog quality         0-15
  investment signals      0-10
  contactability          0-10
  problem clarity         0-10
  ---------------------------------
  total                   0-100

Classificação:
  80-100  -> A  (prioridade alta, candidato à Implantação Marketplace 90)
  65-79   -> B  (prioridade média, candidato à Mentoria ou diagnóstico)
  50-64   -> NUTRIR
  <50     -> NAO_ABORDAR

Este módulo não inventa dados: cada sinal de entrada deve vir de uma fonte
real (site, CNPJ público, marketplace, informação fornecida pelo usuário).
Quando um sinal está ausente, ele é tratado como "desconhecido" e reduz a
confiança do score, nunca é assumido como positivo ou negativo.
"""

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class ScoreInput:
    has_physical_product: Optional[bool] = None
    catalog_size: Optional[int] = None  # número estimado de SKUs
    has_own_brand_or_manufacturer: Optional[bool] = None
    years_active: Optional[int] = None
    has_website: Optional[bool] = None
    has_active_instagram: Optional[bool] = None

    marketplace_presence: dict = field(default_factory=dict)  # {"mercado_livre": bool, "amazon": bool, "shopee": bool}
    marketplace_listing_quality: Optional[str] = None  # "boa" | "fraca" | "ausente" | None
    price_alignment: Optional[str] = None  # "alinhado" | "desalinhado" | None

    public_contact_available: Optional[bool] = None
    decision_maker_identified: Optional[bool] = None

    investment_capacity_signal: Optional[str] = None  # "alto" | "medio" | "baixo" | None
    problem_clarity_signal: Optional[str] = None  # "alto" | "medio" | "baixo" | None


@dataclass
class ScoreResult:
    product_fit: int
    marketplace_gap: int
    business_structure: int
    catalog_quality: int
    investment_signals: int
    contactability: int
    problem_clarity: int
    total: int
    potential: str
    confidence: str
    rationale: str


def _clamp(value: int, low: int, high: int) -> int:
    return max(low, min(high, value))


def compute_score(data: ScoreInput) -> ScoreResult:
    unknowns = 0
    notes: list[str] = []

    # --- product fit (0-20) ---
    product_fit = 0
    if data.has_physical_product is True:
        product_fit += 8
        notes.append("produto físico confirmado")
    elif data.has_physical_product is None:
        unknowns += 1
    else:
        notes.append("sem produto físico identificado")

    if data.catalog_size is not None:
        if 20 <= data.catalog_size <= 500:
            product_fit += 8
            notes.append(f"catálogo de {data.catalog_size} SKUs dentro da faixa ideal")
        elif data.catalog_size > 0:
            product_fit += 3
    else:
        unknowns += 1

    if data.has_own_brand_or_manufacturer is True:
        product_fit += 4
        notes.append("marca própria ou fabricação/distribuição")
    elif data.has_own_brand_or_manufacturer is None:
        unknowns += 1
    product_fit = _clamp(product_fit, 0, 20)

    # --- marketplace gap (0-20): quanto maior a lacuna, maior a pontuação ---
    marketplace_gap = 0
    presence = data.marketplace_presence or {}
    channels_checked = [k for k in ("mercado_livre", "amazon", "shopee") if k in presence]
    if channels_checked:
        present_count = sum(1 for k in channels_checked if presence.get(k))
        absent_count = len(channels_checked) - present_count
        marketplace_gap += absent_count * 5
        notes.append(f"ausente em {absent_count}/{len(channels_checked)} marketplaces verificados")
    else:
        unknowns += 1

    if data.marketplace_listing_quality == "ausente":
        marketplace_gap += 5
    elif data.marketplace_listing_quality == "fraca":
        marketplace_gap += 3
        notes.append("catálogo presente mas com execução fraca")
    elif data.marketplace_listing_quality is None:
        unknowns += 1

    if data.price_alignment == "desalinhado":
        marketplace_gap += 2
    marketplace_gap = _clamp(marketplace_gap, 0, 20)

    # --- business structure (0-15) ---
    business_structure = 0
    if data.years_active is not None:
        if data.years_active >= 3:
            business_structure += 6
            notes.append(f"{data.years_active} anos de operação")
        elif data.years_active > 0:
            business_structure += 2
    else:
        unknowns += 1

    if data.has_website:
        business_structure += 5
    elif data.has_website is None:
        unknowns += 1

    if data.decision_maker_identified:
        business_structure += 4
    business_structure = _clamp(business_structure, 0, 15)

    # --- catalog quality (0-15) ---
    catalog_quality = 0
    if data.catalog_size is not None:
        if 20 <= data.catalog_size <= 500:
            catalog_quality += 8
        elif data.catalog_size > 0:
            catalog_quality += 3
    else:
        unknowns += 1

    if data.has_active_instagram:
        catalog_quality += 4
    elif data.has_active_instagram is None:
        unknowns += 1

    if data.marketplace_listing_quality == "boa":
        catalog_quality += 3
    catalog_quality = _clamp(catalog_quality, 0, 15)

    # --- investment signals (0-10) ---
    investment_signals = {"alto": 10, "medio": 6, "baixo": 2}.get(data.investment_capacity_signal or "", 0)
    if data.investment_capacity_signal is None:
        unknowns += 1

    # --- contactability (0-10) ---
    contactability = 0
    if data.public_contact_available:
        contactability += 6
        notes.append("contato empresarial público disponível")
    elif data.public_contact_available is None:
        unknowns += 1
    if data.decision_maker_identified:
        contactability += 4
    contactability = _clamp(contactability, 0, 10)

    # --- problem clarity (0-10) ---
    problem_clarity = {"alto": 10, "medio": 6, "baixo": 2}.get(data.problem_clarity_signal or "", 0)
    if data.problem_clarity_signal is None:
        unknowns += 1

    total = (
        product_fit
        + marketplace_gap
        + business_structure
        + catalog_quality
        + investment_signals
        + contactability
        + problem_clarity
    )
    total = _clamp(total, 0, 100)

    if total >= 80:
        potential = "A"
    elif total >= 65:
        potential = "B"
    elif total >= 50:
        potential = "NUTRIR"
    else:
        potential = "NAO_ABORDAR"

    if unknowns == 0:
        confidence = "alto"
    elif unknowns <= 3:
        confidence = "medio"
    else:
        confidence = "baixo"

    rationale = "; ".join(notes) if notes else "Poucos sinais confirmados; score conservador."
    if unknowns:
        rationale += f" ({unknowns} sinais desconhecidos não pontuados)."

    return ScoreResult(
        product_fit=product_fit,
        marketplace_gap=marketplace_gap,
        business_structure=business_structure,
        catalog_quality=catalog_quality,
        investment_signals=investment_signals,
        contactability=contactability,
        problem_clarity=problem_clarity,
        total=total,
        potential=potential,
        confidence=confidence,
        rationale=rationale,
    )

File: apps/next/src/lib/score.ts
// Motor de score — porte fiel de packages/agents/score.py (seção 6.3).
// Não inventa dados: sinal ausente conta como desconhecido (0 pontos) e
// reduz a confiança, nunca é assumido como positivo.
import type { Lead, ScoreBreakdown, Potential } from "./types";

const clamp = (v: number, lo: number, hi: number) => Math.max(lo, Math.min(hi, v));

export function computeScore(l: Lead): ScoreBreakdown {
  let unknowns = 0;
  const notes: string[] = [];

  // product fit (0-20)
  let productFit = 0;
  if (l.has_physical_product === true) { productFit += 8; notes.push("produto físico confirmado"); }
  else if (l.has_physical_product == null) unknowns++;
  else notes.push("sem produto físico identificado");
  if (l.catalog_size != null) {
    if (l.catalog_size >= 20 && l.catalog_size <= 500) { productFit += 8; notes.push(`catálogo de ${l.catalog_size} SKUs na faixa ideal`); }
    else if (l.catalog_size > 0) productFit += 3;
  } else unknowns++;
  if (l.has_own_brand === true) { productFit += 4; notes.push("marca própria/fabricação"); }
  else if (l.has_own_brand == null) unknowns++;
  productFit = clamp(productFit, 0, 20);

  // marketplace gap (0-20): quanto maior a lacuna, maior a pontuação
  let gap = 0;
  const presence = l.marketplace_presence ?? {};
  const channels = (["mercado_livre", "amazon", "shopee"] as const).filter((k) => k in presence);
  if (channels.length) {
    const present = channels.filter((k) => presence[k]).length;
    const absent = channels.length - present;
    gap += absent * 5;
    notes.push(`ausente em ${absent}/${channels.length} marketplaces`);
  } else unknowns++;
  if (l.marketplace_quality === "ausente") gap += 5;
  else if (l.marketplace_quality === "fraca") { gap += 3; notes.push("execução fraca no marketplace"); }
  else if (l.marketplace_quality == null) unknowns++;
  gap = clamp(gap, 0, 20);

  // business structure (0-15)
  let biz = 0;
  if (l.years_active != null) {
    if (l.years_active >= 3) { biz += 6; notes.push(`${l.years_active} anos de operação`); }
    else if (l.years_active > 0) biz += 2;
  } else unknowns++;
  if (l.has_website) biz += 5; else if (l.has_website == null) unknowns++;
  if (l.decision_maker_identified) biz += 4;
  biz = clamp(biz, 0, 15);

  // catalog quality (0-15)
  let cat = 0;
  if (l.catalog_size != null) {
    if (l.catalog_size >= 20 && l.catalog_size <= 500) cat += 8;
    else if (l.catalog_size > 0) cat += 3;
  } else unknowns++;
  if (l.has_instagram) cat += 4; else if (l.has_instagram == null) unknowns++;
  if (l.marketplace_quality === "boa") cat += 3;
  cat = clamp(cat, 0, 15);

  // investment signals (0-10)
  const invMap: Record<string, number> = { alto: 10, medio: 6, baixo: 2 };
  const invest = invMap[l.investment_signal ?? ""] ?? 0;
  if (l.investment_signal == null) unknowns++;

  // contactability (0-10)
  let contact = 0;
  if (l.public_contact) { contact += 6; notes.push("contato empresarial público"); }
  else if (l.public_contact == null) unknowns++;
  if (l.decision_maker_identified) contact += 4;
  contact = clamp(contact, 0, 10);

  // problem clarity (0-10)
  const problem = invMap[l.problem_signal ?? ""] ?? 0;
  if (l.problem_signal == null) unknowns++;

  let total = clamp(productFit + gap + biz + cat + invest + contact + problem, 0, 100);

  let potential: Potential;
  if (total >= 80) potential = "A";
  else if (total >= 65) potential = "B";
  else if (total >= 50) potential = "NUTRIR";
  else potential = "NAO_ABORDAR";

  const confidence = unknowns === 0 ? "alto" : unknowns <= 3 ? "medio" : "baixo";
  let rationale = notes.length ? notes.join("; ") : "Poucos sinais confirmados; score conservador.";
  if (unknowns) rationale += ` (${unknowns} sinais desconhecidos não pontuados).`;

  return {
    product_fit: productFit,
    marketplace_gap: gap,
    business_structure: biz,
    catalog_quality: cat,
    investment_signals: invest,
    contactability: contact,
    problem_clarity: problem,
    total,
    potential,
    confidence,
    rationale,
  };
}
